### ISSUE TITLE
GUS pop elimination for hardware mixed sources / Star Control 2

### ISSUE THREAD

**Issue by marsej** [dosbox-staging/dosbox-staging#3350]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

v.0.81.0-rc2

### What Operating System are you using?

Windows 10

### Is your feature request related to a problem? Please describe.

Star Control 2 has pop sound issues in 7 songs when using GUS. I'm not sure if the pop happens at end of some samples or what but if so..

I read from https://www.vogons.org/viewtopic.php?t=36661 that Star Control 2 uses hardware GUS mixing.

That means that the GUS hardware mixing code should see the samples individually and some sort of fade could be added there as option.

Previously I was using dosbox-staging-windows-msys2-x86_64-v0.79.0-alpha-166-gb6083 and the popping was not as evident with it as later versions but to my ears the sound quality is slightly better with later builds. (attached file EncounterPops has comparisons between dosbox staging versions)

### Describe the solution you'd like

Perhaps an optional pop eliminating/fading/filtering mechanism could be added if the case is that there's some sort of abrupt end to a sample detected coming to the hardware mixer.

### Add any other context or screenshots about the feature request here.

The "encounter" is called redalert.mod and does not have pop in cmod which also uses gus hardware mixing (says a list I found). So we know the issue is in the Star Control 2 code. However, the talkingpet.mod also has pops in cmod.

[EncounterPops.zip](https://github.com/dosbox-staging/dosbox-staging/files/14047842/EncounterPops.zip)
[sc2popping.zip](https://github.com/dosbox-staging/dosbox-staging/files/14047844/sc2popping.zip)

In-game the redalert.com/encounter music plays on every time alien ship is met, so just load (almost) any of the saves attached to hear it almost immediately.

[SC2MUSICSAVES.zip](https://github.com/dosbox-staging/dosbox-staging/files/14047883/SC2MUSICSAVES.zip)

I used SC2 CD version with irq 5 and dma 5. (CD ver allows to skip the star map lookup)

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by marsej** [comment]
Here's CMOD with the mods (redalert, talkingpet).
[CMODwithSC2Mods.zip](https://github.com/dosbox-staging/dosbox-staging/files/14047930/CMODwithSC2Mods.zip)

**Comment by johnnovak** [comment]
@marsej I assume you read my comments and general thought about this here:
https://github.com/dosbox-staging/dosbox-staging/pull/3310#issuecomment-1909168326

We might implement an enhancement to minimise/get rid of the pops, but I'm pretty sure that's how the game behaves on real hardware too.

**Comment by marsej** [comment]
Another thing to consider: I don't believe that proper CRT emulation involves close-up comparisons like most in the thread linked above. To get accurate emulation you have to take photos at normal viewing distance and then perhaps use artistic abilities to match the emulation to those photos, rather than emulating scanlines etc that are visible only at close distance.

**Comment by marsej** [comment]
Also think that despite all the odd color anomalies, the one on the left is bit more scary and immersive. Also this is not enough to me: There's no PC monitors with plain glass on top of the pixels like with CRT's. Any light that hits the matte monitors plastic causes a plastic sheen that I argue reduces immersion. So to get proper emulation you need Apple Studio Display or OLED TV.

![crt vs lcd](https://github.com/dosbox-staging/dosbox-staging/assets/81806010/6414c127-b24b-4f7d-9314-1df1197d9114)

**Comment by Grounded0** [comment]
MacBooks also have glass screens.

**Comment by Grounded0** [comment]
I always had CRTs nicer than usual and they always aspired for very neutral image quality and thus i aspire for a neutral image quality myself so i have a HiDPI display and use `sharp` shader. Looking at a CRT that shitty would've made my life miserable because it perverts what is actually trying to be presented.

**Comment by johnnovak** [comment]
@marsej I have 9 CRTs here currently, and I've tweaked the shaders so they're pretty accurate 😎 As @Grounded0 say, they pretty much look perfect on a glass 4K Macbook Pro display.

> Most pixel art was drawn on a large high quality displays using Deluxe Paint heavily zoomed in so you're actually perverting the original artistic intention. 😁

No, man, far from it 😄 Most people used Amigas to pixel because DPaint II on the PC is a bit shit. So yeah, even for PC games up until the late 90s until low-res fixed-palette pixel art was relevant. And yeah, they usually used 14" 15kHz screens. In DPaint you always had the 1:1 image on the side as shown below, otherwise pixeling is pointless... I've done quite a bit of pixeling back in the day.

This is how you typically do it. In fact, you can't even _not_ show the 100% image in DPaint 😄 

https://blog.johnnovak.net/2022/04/15/achieving-period-correct-graphics-in-personal-computer-emulators-part-1-the-amiga/img/guru5-dpaint3.png

**Comment by johnnovak** [comment]
@marsej Your example images are consoles hooked to up to TVs via composite. PC CRT monitors are sharp as nails. My late 90s VGAs I have are still tack sharp...

**Comment by marsej** [comment]
Does late 90s count for dosbox tho? I'd like to have (working) a large dot pitch shadow mask CRT for normal doing view distance photos and comparison. Ultimately however regardless of what the CRTs actually show, there's a case for upscaling. ie. just like when you consume SD content with 4k TV, you need Panasonic or Sony TV with tuned algos to upscale it for most pleasing results. I only remember that when my CRT broke and I switched to LCD, games suddenly looked so much less immersive. So my goal is "more immersive" and not some exact emulation of scanlines which I don't remember. I get bothered by too visible scanlines so I find it hard to believe that is what we should be aiming for.

**Comment by marsej** [comment]
Regardless, having "more immersive" as an option in the settings is the same as saying end users desire pleasing SD upscaling. So I have the lottes shader but that has never "shipped" with dosbox right? I only found it in another repository and it needed tweaking. Possible futur goal: Ship an immersive "upscaling" shader with dosbox.

**Comment by johnnovak** [comment]
Use `crt-auto-arcade`, that has some basis in reality. What you call immersive emulates a really really bad TV 😄 But no TV I've ever seen has such strong "blur"... So that's out of scope, but the good news is, you can keep using such things as custom shaders. The option will always be there.

**Comment by marsej** [comment]
The "blur" is supposed to mimic the glow of brightest phoshor onto the neighbouring area, as seen by eye (not necessarily captured by camera so well or easily, just like sun isn't).

**Comment by marsej** [comment]
![glow](https://github.com/dosbox-staging/dosbox-staging/assets/81806010/5f38b390-71d7-4893-bd75-0e24b24def2a)

**Comment by Grounded0** [comment]
> S. Here's the source for the image: https://twitter.com/CRTpixels/status/1389639277752033282
The tv is said to be Sanyo DS13320 13" CRT .. someone is selling one on ebay US right now.

That's a TV display not a monitor. They call them monitors exactly for the same reason we call high-quality neutral sounding speakers monitors. They just reproduce the image in as authentic way as possible hence they look fairly similar to the source image.

**Comment by marsej** [comment]
Well "arcade monitor" is a consumption device (see comment below). Am I or Timothy Lottes just imagining things or was the glow significant part of how things looked on some of these "consumption monitors". I find it much more important than the masks, lines and so forth. Anything that hides the jaggy pixels for me. High end for production/mixing, low end high dot pitch glowwy for consuming low res content.

```
//   by Timothy Lottes
//
// This is more along the style of a really good CGA arcade monitor.
// With RGB inputs instead of NTSC.
// The shadow mask example has the mask rotated 90 degrees for less chromatic aberration.
//
// Left it unoptimized to show the theory behind the algorithm.
//
// **It is an example what I personally would want as a display option for pixel art games.**
// Please take and use, change, or whatever.
```

** To emphasize that I am not the only one out there desiring things to look better than on high-end production focused monitors.

**Comment by kcgen** [comment]
Thanks for opening a ticket about this GUS issue (re-pasting my comment from the DC bias PR):

I rook a recording here from SC2 launched w/ the cracked .com file.

https://github.com/dosbox-staging/dosbox-staging/assets/1557255/9712e7f7-254e-43bc-8ae7-ed93a11399b8

Note that at 0:28 and 0:33 the oscilloscope signal's mid-line bumps upward corresponding to a small audible pop.

The same talking pet MOD played in the [Ur-Quan Masters](https://en.wikipedia.org/wiki/The_Ur-Quan_Masters) implementation also generates the same pops (also at the 0.28 and 0:33 marks), seen in this YouTube video:

https://www.youtube.com/watch?app=desktop&v=MrvxgcFCyKw

This comparison eliminates (all) DOSBox(es) and even the GUS (both hardware and software) as the source of this problem (because the UQM implementation plays PCM directly to the host audio device).

So my hunch is the talking pet MOD file needs all its internal tracks normalized/centred; maybe the UQM community has fixed it?(perhaps just for UQM, but ideally the fixed mode could be hacked it back into the original DOS data files).

Dealing with this additive DC bias is quite a bit harder than it is in the OPL or PC speaker (even if we can isolate voices within the GUS). This also isn't a habitual problem - but isolated to specific MODs (and most of them in Star Control 2, unfortunately).

So I don't think trying to fix this at the GUS emulation level - because it's not the source of the problem. It should be tacked in the MODs themselves.

**Comment by johnnovak** [comment]
TL;DR summary:

- The game composer was really sloppy with th samples; they contain DC offsets and no fade-ins and fade-outs.
- The GUS MOD player of the game is not very good; it's one of those "naive" clicky GUS playroutines.
- There's nothing for us to do.

**Comment by johnnovak** [comment]
> Well "arcade monitor" is a consumption device (see comment below). Am I or Timothy Lottes just imagining things or was the glow significant part of how things looked on some of these "consumption monitors".

Timothy was right, CGA monitors are a blurry mess and everybody hated them. Then you went and edited the shader to make the blur way worse... 😅 

EGA and VGA monitors fixed CGA's mistakes and were tack sharp. Gaming on PC was an afterthought, the whole PC monitor tech was optimised for nothing else but text legibility and sharpness so you can stare at spreadsheets all day. 

But not even the worst CGA monitor on the planet had such extreme blur.

You like it? Fine.
Is it authentic? Not even remotely.
Will such a preset be a bundled preset? No way! 😅
